xgi/classes/function.py

```diff
@@ -688,3 +690,82 @@ def convert_labels_to_integers(H, label_attribute="label"):
     )
 
     return temp_H
+
+
+def density(H, order=None, max_order=None, ignore_singletons=False):
+    r"""Hypergraph density.
+
+    The density of a hypergraph is the number of existing edges divided by the number of
+    possible edges.
+
+    Let `H` have :math:`n` nodes and :math:`m` hyperedges. Then,
+
+    * `density(H) =` :math:`\frac{m}{2^n - 1}`,
+    * `density(H, count_singletons=True) =` :math:`\frac{m}{2^n - 1 - n}`.
+
+    Here, :math:`2^n` is the total possible number of hyperedges on `H`, to which we
+    subtract :math:`1` because the empty hyperedge is not considered, or :math:`n` when
+    not considering singletons.
+
+    Now assume `H` has :math:`a` edges with order :math:`1` and :math:`b` edges with
+    order :math:`2`.  Then,
+
+    * `density(H, order=1) =` :math:`\frac{a}{{n \choose 2}}`,
+    * `density(H, order=2) =` :math:`\frac{b}{{n \choose 3}}`,
+    * `density(H, max_order=1) =` :math:`\frac{a}{{n \choose 1} + {n \choose 2}}`,
+    * `density(H, max_order=1, ignore_singletons=True) =` :math:`\frac{a}{{n \choose 2}}`,
+    * `density(H, max_order=2) =` :math:`\frac{m}{{n \choose 1} + {n \choose 2} + {n \choose 3}}`,
+    * `density(H, max_order=2, ignore_singletons=True) =` :math:`\frac{m}{{n \choose 2} + {n \choose 3}}`,
+
+    Parameters
+    ---------
+    order : int or None (default)
+        If not None, only count edges of the specified order.
+
+    max_order : int or None (default)
+        If not None, only count edges of order up to this value, inclusive.
+
+    ignore_singletons : bool (default False)
+        Whether to consider singleton edges.  Ignored if `order` is not None and
+        different from :math:`0`.
+
+    Notes
+    -----
+    If both `order` and `max_order` are not None, `max_order` is ignored.
+
+    """
+    n = H.num_nodes
+    if n < 1 or H.num_edges < 1:
```

Comment:
Do we want to return 0 if there are 0 nodes? The denom would be 0 technically.
Could also return None or some error like "Cannot compute the density of empty Hypergraph".

Reply:
Good catch!